src/compiler/factory.ts

```diff
@@ -2181,51 +2181,120 @@ namespace ts {
 
     // JSDoc
 
-    /* @internal */
+    function createJSDocTag<T extends JSDocTag>(kind: T["kind"], tagName: string, comment?: string): T {
+        const node = createSynthesizedNode(kind) as T;
+        node.tagName = createIdentifier(tagName);
+        node.comment = comment;
+        return node;
+    }
+
+    function createJSDocTagWithTypeExpression<T extends JSDocTag & { typeExpression?: TypeNode }>(kind: T["kind"], tagName: string, typeExpression?: TypeNode, comment?: string) {
+        const node = createJSDocTag<T>(kind, tagName, comment);
+        node.typeExpression = typeExpression;

```

Comment:
AFAIK the factory functions should set properties in the same order as the parser to avoid multiple shapes of the same object.
Looking at `parseReturnTag` as an example: it sets `tagName`, `typeExpression` and `comment` (in `parseTag` function).

Maybe you want to change the parser to initialize the properties in the same order as the newly added factories (`comment` before `typeExpression` and so on)

Reply:
In fact, I also have concerns about this, I chose to remove this helper function.😢